lib/galaxy/tools/deps/resolvers/galaxy_packages.py

```diff
@@ -25,40 +31,68 @@ def resolve( self, name, version, type, **kwds ):
         symbolic link (if found). Returns a triple of: env_script, base_path, real_version
         """
         if version is None or self.versionless:
-            return self._find_dep_default( name, type=type, **kwds )
+            exact = not self.versionless or version is None
+            return self._find_dep_default( name, type=type, exact=exact, **kwds )
         else:
             return self._find_dep_versioned( name, version, type=type, **kwds )
 
     def _find_dep_versioned( self, name, version, type='package', **kwds ):
         base_path = self.base_path
         path = join( base_path, name, version )
-        return self._galaxy_package_dep(path, version)
+        return self._galaxy_package_dep(path, version, True)
 
-    def _find_dep_default( self, name, type='package', **kwds ):
+    def _find_dep_default( self, name, type='package', exact=True, **kwds ):
         base_path = self.base_path
         path = join( base_path, name, 'default' )
         if islink( path ):
             real_path = realpath( path )
             real_version = basename( real_path )
-            return self._galaxy_package_dep(real_path, real_version)
+            return self._galaxy_package_dep(real_path, real_version, exact)
         else:
             return INDETERMINATE_DEPENDENCY
 
-    def _galaxy_package_dep( self, path, version ):
+    def _galaxy_package_dep( self, path, version, exact ):
         script = join( path, 'env.sh' )
         if exists( script ):
-            return GalaxyPackageDependency(script, path, version)
+            return GalaxyPackageDependency(script, path, version, exact)
         elif exists( join( path, 'bin' ) ):
-            return GalaxyPackageDependency(None, path, version)
+            return GalaxyPackageDependency(None, path, version, exact)
         return INDETERMINATE_DEPENDENCY
 
 
+class GalaxyPackageDependencyResolver(BaseGalaxyPackageDependencyResolver, ListableDependencyResolver):
+    resolver_type = "galaxy_packages"
+
+    def list_dependencies(self):
+        base_path = self.base_path
+        for package_name in listdir(base_path):
+            package_dir = join(base_path, package_name)
+            for version in listdir(package_dir):
+                version_dir = join(package_dir, version)
+                if version == "default":
+                    version = None
+                valid_dependency = _is_dependency_directory(version_dir)
+                if valid_dependency:
+                    yield self._to_requirement(package_name, version)
+
+
+def _is_dependency_directory(directory):
+    return exists(join(directory, 'env.sh')) or exists(join(directory, 'bin'))
```

Comment:
I was not talking about `versioned packages`, this was more a general comment. I didn't know that a `tool_deps`/`package_name`/`random`/`bin` will be picked up as dependency. Great if it does!